algorithms/src/snark/marlin/ahp/prover/round_functions/third.rs

```diff
@@ -21,249 +20,286 @@ use crate::{
-                    largest_non_zero_domain_size,
-                    &circuit.fft_precomputation,
-                    &circuit.ifft_precomputation,
-                );
-                (*circuit, result)
-            });
-
-            let label_g_b = witness_label(circuit.id, "g_b", 0);
-            pool.add_job(move || {
-                let result = Self::matrix_sumcheck_helper(
-                    label_g_b,
-                    circuit_state.non_zero_b_domain,
-                    &circuit.b_arith,
-                    *alpha,
-                    beta,
-                    v_H_i_alpha_v_H_i_beta,
-                    largest_non_zero_domain_size,
-                    &circuit.fft_precomputation,
-                    &circuit.ifft_precomputation,
-                );
-                (*circuit, result)
-            });
-
-            let label_g_c = witness_label(circuit.id, "g_c", 0);
-            pool.add_job(move || {
-                let result = Self::matrix_sumcheck_helper(
-                    label_g_c,
-                    circuit_state.non_zero_c_domain,
-                    &circuit.c_arith,
-                    *alpha,
-                    beta,
-                    v_H_i_alpha_v_H_i_beta,
-                    largest_non_zero_domain_size,
-                    &circuit.fft_precomputation,
-                    &circuit.ifft_precomputation,
-                );
-                (*circuit, result)
-            });
-        }
+        let zk_bound = Self::zk_bound();
 
-        let mut sums = Vec::with_capacity(state.circuit_specific_states.len());
-        let mut gs = BTreeMap::new();
-        for ((circuit_a, (sum_a, lhs_a, g_a)), (circuit_b, (sum_b, lhs_b, g_b)), (circuit_c, (sum_c, lhs_c, g_c))) in
-            pool.execute_all().into_iter().tuples()
-        {
-            assert_eq!(circuit_a, circuit_b);
-            assert_eq!(circuit_a, circuit_c);
-            let matrix_sum = prover::message::MatrixSums { sum_a, sum_b, sum_c };
-            sums.push(matrix_sum);
-            state.circuit_specific_states.get_mut(circuit_a).unwrap().lhs_polynomials = Some([lhs_a, lhs_b, lhs_c]);
-            let matrix_gs = prover::MatrixGs { g_a, g_b, g_c };
-            gs.insert(circuit_a.id, matrix_gs);
-        }
+        let max_variable_domain = state.max_variable_domain;
+
+        let verifier::FirstMessage { batch_combiners } = verifier_message;
+        let verifier::SecondMessage { alpha, eta_b, eta_c } = verifier_second_message;
 
-        let msg = prover::ThirdMessage { sums };
-        let oracles = prover::ThirdOracles { gs };
+        let (h_1, x_g_1_sum, msg) = Self::calculate_lhs(&mut state, batch_combiners, alpha, eta_b, eta_c)?;
 
-        assert!(oracles.matches_info(&Self::third_round_polynomial_info(
-            state.circuit_specific_states.keys().map(|c| (c.id, &c.index_info))
-        )));
+        let mut sumcheck_lhs = h_1.mul_by_vanishing_poly(max_variable_domain);
+        sumcheck_lhs += &x_g_1_sum;
```

Comment:
Let's put this also inside the `#[cfg(debug_assertions)]`

Reply:
dcb2264bb